protostar/starknet_gateway/gateway_facade.py

```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
         except KeyError:
             raise UnknownFunctionException(function_name) from KeyError
 
+    async def _get_resolved_contract_function(
+        self,
+        contract_address: AddressRepresentation,
+        client: Optional[Client] = None,
+    ):
+        try:
+            try:
+                contract = await Contract.from_address(
+                    address=contract_address,
+                    client=client or self._gateway_client,
+                    proxy_config=True,
+                )
+                return contract
+            except ClientError as err:
+                if "not deployed" in err.message:
+                    raise ContractNotFoundException(
+                        contract_address=contract_address
+                    ) from err
+                raise err
+        except ProxyResolutionError:
+            contract = await Contract.from_address(
+                address=contract_address,
+                client=client or self._gateway_client,
+                proxy_config=False,
+            )
+            return contract
```

Comment:
This logic is pretty tangled up - i'd break that down into small methods. Moreover, can we explain a bit why we do things that way? Or make code do that itself instead?